src/bson.rs

```diff
@@ -304,59 +304,108 @@ impl From<Value> for Bson {
+            ),
             Value::Null => Bson::Null,
         }
     }
 }
 
 impl From<Bson> for Value {
     fn from(bson: Bson) -> Self {
-        match bson {
+        bson.into_relaxed_extjson()
+    }
+}
+
+impl Bson {
+    pub fn into_relaxed_extjson(self) -> Value {
+        match self {
+            Bson::FloatingPoint(v) if v.is_nan() => {
+                let s = if v.is_sign_negative() { "-NaN" } else { "NaN" };
+
+                json!({ "$numberDouble": s })
+            }
+            Bson::FloatingPoint(v) if v.is_infinite() => {
+                let s = if v.is_sign_negative() {
+                    "-Infinity"
+                } else {
+                    "Infinity"
+                };
+
+                json!({ "$numberDouble": s })
+            }
             Bson::FloatingPoint(v) => json!(v),
             Bson::String(v) => json!(v),
             Bson::Array(v) => json!(v),
-            Bson::Document(v) => json!(v),
+            Bson::Document(v) => {
+                Value::Object(v.into_iter().map(|(k, v)| (k, Value::from(v))).collect())
+            }
             Bson::Boolean(v) => json!(v),
             Bson::Null => Value::Null,
-            Bson::Regex(Regex { pattern, options }) => json!({
-                "$regex": pattern,
-                "$options": options
-            }),
+            Bson::Regex(Regex { pattern, options }) => {
+                let mut chars: Vec<_> = options.chars().collect();
+                chars.sort();
+
+                let options: String = chars.into_iter().collect();
+
+                json!({
+                    "$regularExpression": {
+                        "pattern": pattern,
+                        "options": options,
+                    }
+                })
+            }
             Bson::JavaScriptCode(code) => json!({ "$code": code }),
             Bson::JavaScriptCodeWithScope(JavaScriptCodeWithScope { code, scope }) => json!({
                 "$code": code,
-                "scope": scope
+                "$scope": scope,
             }),
             Bson::I32(v) => v.into(),
             Bson::I64(v) => v.into(),
             Bson::TimeStamp(TimeStamp { time, increment }) => json!({
-                "t": time,
-                "i": increment
+                "$timestamp": {
+                    "t": time,
+                    "i": increment,
+                }
             }),
             Bson::Binary(Binary { subtype, ref bytes }) => {
                 let tval: u8 = From::from(subtype);
                 json!({
-                    "type": tval,
-                    "$binary": hex::encode(bytes),
+                    "$binary": {
+                        "base64": base64::encode(bytes),
+                        "subType": hex::encode([tval.into()]),
+                    }
+                })
+            }
+            Bson::ObjectId(v) => json!({"$oid": v.to_hex()}),
+            Bson::UtcDatetime(v) if v.timestamp_millis() >= 0 && v.year() <= 99999 => {
```

Comment:
For relaxed extended JSON, the spec requires that dates from `[unix epoch, Y10000)` are formatted as strings rather than specified as integers since the epoch.